File: go/engine/rules.go
package engine

// =============================================================================
// MOVE TYPES
// =============================================================================

// Placement represents placing a single tile at a position.
//
// A move consists of one or more placements, all in the same row or column.
type Placement struct {
	Pos  Position // Where to place the tile
	Tile Tile     // The tile being placed
}

// Move represents a complete player move.
//
// A valid move in Qwirkle:
//   - Places 1-6 tiles from hand
//   - All tiles in same row OR same column
//   - Tiles form a continuous line (possibly filling gaps)
//   - Each formed line is valid (same color or same shape, no duplicates)
type Move struct {
	Placements []Placement // The tiles and positions
	Score      int         // Points this move would earn
}

// =============================================================================
// LINE EXTRACTION
// =============================================================================

// GetLine returns all tiles in a line starting from pos, moving in direction (dr, dc).
//
// Parameters:
//   - board: The game board
//   - pos: Starting position (not included in result)
//   - dr: Row delta (-1=up, 0=stay, 1=down)
//   - dc: Column delta (-1=left, 0=stay, 1=right)
//
// Returns tiles in order from pos outward.
//
// Example: GetLine(board, {0,0}, 0, 1) returns tiles at (0,1), (0,2), ... until empty.
func GetLine(board *Board, pos Position, dr, dc int) []Tile {
	tiles := make([]Tile, 0)

	// Start one step in the direction
	r, c := pos.Row+dr, pos.Col+dc

	// Keep going while we find tiles
	for {
		p := Position{Row: r, Col: c}
		if t := board.Get(p); t != nil {
			tiles = append(tiles, *t)
			r += dr // Move to next position
			c += dc
		} else {
			break // Empty cell - end of line
		}
	}

	return tiles
}

// GetHorizontalLine returns all tiles in the horizontal line through pos.
//
// Includes the tile at pos (if any) and extends left and right until empty cells.
// Used for scoring - we need to count the complete line a tile joins.
func GetHorizontalLine(board *Board, pos Position) []Tile {
	// Get tiles to the left (will be in reverse order)
	left := GetLine(board, pos, 0, -1)

	// Get tiles to the right
	right := GetLine(board, pos, 0, 1)

	// Build result: reversed left + center + right
	// Pre-allocate for efficiency
	result := make([]Tile, 0, len(left)+len(right)+1)

	// Reverse the left tiles (they were collected going away from pos)
	for i := len(left) - 1; i >= 0; i-- {
		result = append(result, left[i])
	}

	// Add center tile if present
	if t := board.Get(pos); t != nil {
		result = append(result, *t)
	}

	// Add right tiles
	result = append(result, right...)

	return result
}

// GetVerticalLine returns all tiles in the vertical line through pos.
func GetVerticalLine(board *Board, pos Position) []Tile {
	up := GetLine(board, pos, -1, 0)
	down := GetLine(board, pos, 1, 0)

	result := make([]Tile, 0, len(up)+len(down)+1)

	// Reverse up tiles
	for i := len(up) - 1; i >= 0; i-- {
		result = append(result, up[i])
	}

	if t := board.Get(pos); t != nil {
		result = append(result, *t)
	}

	result = append(result, down...)

	return result
}

// =============================================================================
// OPTIMIZED LINE EXTRACTION (zero allocation)
// =============================================================================

// LineTiles is a fixed-size buffer for line extraction (max 6 tiles + 1 center).
// Using a fixed array avoids heap allocations in hot paths.
type LineTiles struct {
	tiles [7]Tile
	count int
}

// Len returns the number of tiles in the line.
func (lt *LineTiles) Len() int {
	return lt.count
}

// Slice returns a slice view of the tiles (no allocation).
func (lt *LineTiles) Slice() []Tile {
	return lt.tiles[:lt.count]
}

// GetHorizontalLineFast extracts horizontal line into provided buffer.
// Returns the number of tiles found. No heap allocation.
func GetHorizontalLineFast(board *Board, pos Position, buf *LineTiles) {
	buf.count = 0

	// Scan left - count tiles
	leftCount := 0
	for c := pos.Col - 1; ; c-- {
		if board.Get(Position{Row: pos.Row, Col: c}) != nil {
			leftCount++
		} else {
			break
		}
	}

	// Add left tiles in correct order
	for i := 0; i < leftCount && buf.count < 7; i++ {
		c := pos.Col - leftCount + i
		t := board.Get(Position{Row: pos.Row, Col: c})
		buf.tiles[buf.count] = *t
		buf.count++
	}

	// Add center tile
	if buf.count < 7 {
		if t := board.Get(pos); t != nil {
			buf.tiles[buf.count] = *t
			buf.count++
		}
	}

	// Scan right
	for c := pos.Col + 1; buf.count < 7; c++ {
		if t := board.Get(Position{Row: pos.Row, Col: c}); t != nil {
			buf.tiles[buf.count] = *t
			buf.count++
		} else {
			break
		}
	}
}

// GetVerticalLineFast extracts vertical line into provided buffer.
// Returns the number of tiles found. No heap allocation.
func GetVerticalLineFast(board *Board, pos Position, buf *LineTiles) {
	buf.count = 0

	// Scan up - count tiles
	upCount := 0
	for r := pos.Row - 1; ; r-- {
		if board.Get(Position{Row: r, Col: pos.Col}) != nil {
			upCount++
		} else {
			break
		}
	}

	// Add up tiles in correct order
	for i := 0; i < upCount && buf.count < 7; i++ {
		r := pos.Row - upCount + i
		t := board.Get(Position{Row: r, Col: pos.Col})
		buf.tiles[buf.count] = *t
		buf.count++
	}

	// Add center tile
	if buf.count < 7 {
		if t := board.Get(pos); t != nil {
			buf.tiles[buf.count] = *t
			buf.count++
		}
	}

	// Scan down
	for r := pos.Row + 1; buf.count < 7; r++ {
		if t := board.Get(Position{Row: r, Col: pos.Col}); t != nil {
			buf.tiles[buf.count] = *t
			buf.count++
		} else {
			break
		}
	}
}

// IsValidLineFast checks line validity using the LineTiles buffer.
func IsValidLineFast(lt *LineTiles) bool {
	if lt.count <= 1 {
		return true
	}
	if lt.count > 6 {
		return false
	}

	tiles := lt.tiles[:lt.count]

	// Check for duplicates using fixed array
	var seen [36]bool
	for _, t := range tiles {
		idx := t.Index()
		if seen[idx] {
			return false
		}
		seen[idx] = true
	}

	// Check same color
	firstColor := tiles[0].Color
	sameColor := true
	for i := 1; i < lt.count; i++ {
		if tiles[i].Color != firstColor {
			sameColor = false
			break
		}
	}
	if sameColor {
		return true
	}

	// Check same shape
	firstShape := tiles[0].Shape
	for i := 1; i < lt.count; i++ {
		if tiles[i].Shape != firstShape {
			return false
		}
	}
	return true
}

// =============================================================================
// LINE VALIDATION
// =============================================================================

// IsValidLine checks if a line of tiles follows Qwirkle rules.
//
// Rules for a valid line:
// 1. All tiles share the same color (with different shapes), OR
// 2. All tiles share the same shape (with different colors)
// 3. No duplicate tiles (same shape AND color)
// 4. Maximum 6 tiles (one of each in the shared attribute)
//
// OPTIMIZED: Uses fixed-size array instead of map for duplicate checking.
func IsValidLine(tiles []Tile) bool {
	// Empty or single tile is always valid
	if len(tiles) <= 1 {
		return true
	}

	// Maximum line length is 6 (one of each in a set)
	if len(tiles) > 6 {
		return false
	}

	// Check for duplicates using fixed array (no allocation)
	// 36 unique tiles: 6 shapes × 6 colors
	var seen [36]bool
	for _, t := range tiles {
		idx := t.Index()
		if seen[idx] {
			return false // Duplicate found
		}
		seen[idx] = true
	}

	// Check if all same color
	sameColor := true
	firstColor := tiles[0].Color
	for i := 1; i < len(tiles); i++ {
		if tiles[i].Color != firstColor {
			sameColor = false
			break
		}
	}

	if sameColor {
		return true
	}

	// Check if all same shape
	firstShape := tiles[0].Shape
	for i := 1; i < len(tiles); i++ {
		if tiles[i].Shape != firstShape {
			return false
		}
	}
	return true
}

// =============================================================================
// PLACEMENT VALIDATION
// =============================================================================

// ValidatePlacement checks if placing a tile at pos is valid.
//
// This temporarily places the tile to check both lines it would form.
// Caller should ensure the position is empty before calling.
//
// Parameters:
//   - board: Current board state (may be modified temporarily)
//   - pos: Where to place the tile
//   - tile: The tile to place
//   - isFirstMove: True if this is the first tile of the game
//
// Note: This function temporarily modifies the board but restores it.
// We use defer to ensure cleanup even if something panics.
func ValidatePlacement(board *Board, pos Position, tile Tile, isFirstMove bool) bool {
	// Position must be empty
	if board.Has(pos) {
		return false
	}

	// First tile of game must be at origin (0,0)
	if isFirstMove && board.IsEmpty() {
		return pos.Row == 0 && pos.Col == 0
	}

	// After first move, tile must connect to existing tiles
	if !isFirstMove && !board.HasNeighbor(pos) {
		return false
	}

	// Temporarily place tile to check lines
	board.Set(pos, tile)
	// defer ensures Remove is called when function returns
	// This is Go's way of doing cleanup (like try-finally in Java)
	defer board.Remove(pos)

	// Check horizontal line validity
	hLine := GetHorizontalLine(board, pos)
	if !IsValidLine(hLine) {
		return false
	}

	// Check vertical line validity
	vLine := GetVerticalLine(board, pos)
	if !IsValidLine(vLine) {
		return false
	}

	return true
}

// ValidateSingleTile is a fast path for validating single-tile placements.
// Avoids board cloning and heap allocations. Optimized for Monte Carlo.
func ValidateSingleTile(board *Board, pos Position, tile Tile) bool {
	// Position must be empty
	if board.Has(pos) {
		return false
	}

	// First tile must go at origin
	if board.IsEmpty() {
		return pos.Row == 0 && pos.Col == 0
	}

	// Must connect to existing tiles
	if !board.HasNeighbor(pos) {
		return false
	}

	// Check lines using zero-allocation buffers
	board.Set(pos, tile)

	var hBuf, vBuf LineTiles
	GetHorizontalLineFast(board, pos, &hBuf)
	if !IsValidLineFast(&hBuf) {
		board.Remove(pos)
		return false
	}

	GetVerticalLineFast(board, pos, &vBuf)
	valid := IsValidLineFast(&vBuf)

	board.Remove(pos)
	return valid
}

// ValidateMove checks if a complete move (multiple placements) is valid.
//
// Move validity rules:
// 1. At least one tile must be placed
// 2. All tiles must be in same row OR same column
// 3. Tiles must form a contiguous line (can fill gaps between existing tiles)
// 4. Each individual placement must be valid
func ValidateMove(board *Board, placements []Placement, isFirstMove bool) bool {
	// Fast path for single tile placement
	if len(placements) == 1 {
		return ValidateSingleTile(board, placements[0].Pos, placements[0].Tile)
	}
	if len(placements) == 0 {
		return false
	}

	// Check collinearity: all tiles in same row OR same column
	allSameRow := true
	allSameCol := true
	for i := 1; i < len(placements); i++ {
		if placements[i].Pos.Row != placements[0].Pos.Row {
			allSameRow = false
		}
		if placements[i].Pos.Col != placements[0].Pos.Col {
			allSameCol = false
		}
	}

	// Must be collinear (or single tile, which is trivially both)
	if !allSameRow && !allSameCol {
		return false
	}

	// Clone board to test placements
	// We need to apply them one by one and validate
	testBoard := board.Clone()

	for i, p := range placements {
		// First placement of first move gets special handling
		isFirst := isFirstMove && i == 0 && board.IsEmpty()
		if !ValidatePlacement(testBoard, p.Pos, p.Tile, isFirst) {
			return false
		}
		// Apply this placement before validating next
		testBoard.Set(p.Pos, p.Tile)
	}

	// Check contiguity: all placements must form a continuous line
	// This means no empty gaps between the placed tiles
	if len(placements) > 1 {
		if allSameRow {
			// All in same row - check column continuity
			row := placements[0].Pos.Row

			// Find min and max columns
			minCol, maxCol := placements[0].Pos.Col, placements[0].Pos.Col
			for _, p := range placements[1:] {
				if p.Pos.Col < minCol {
					minCol = p.Pos.Col
				}
				if p.Pos.Col > maxCol {
					maxCol = p.Pos.Col
				}
			}

			// Every position between min and max must be filled
			// (either by existing tile or new placement)
			for c := minCol; c <= maxCol; c++ {
				if !testBoard.Has(Position{Row: row, Col: c}) {
					return false // Gap found
				}
			}
		} else { // allSameCol
			// All in same column - check row continuity
			col := placements[0].Pos.Col

			minRow, maxRow := placements[0].Pos.Row, placements[0].Pos.Row
			for _, p := range placements[1:] {
				if p.Pos.Row < minRow {
					minRow = p.Pos.Row
				}
				if p.Pos.Row > maxRow {
					maxRow = p.Pos.Row
				}
			}

			for r := minRow; r <= maxRow; r++ {
				if !testBoard.Has(Position{Row: r, Col: col}) {
					return false
				}
			}
		}
	}

	return true
}

// =============================================================================
// SCORING
// =============================================================================

// ScoreMove calculates points for a move.
//
// Scoring rules:
//   - Each tile in a completed line scores 1 point
//   - A tile can be part of two lines (horizontal and vertical)
//   - A "Qwirkle" (6-tile line) scores 6 bonus points
//   - Placing a single tile with no neighbors scores 1 point
//
// IMPORTANT: Call this AFTER placements are applied to the board.
// The board should have the new tiles in place.
func ScoreMove(board *Board, placements []Placement) int {
	if len(placements) == 0 {
		return 0
	}

	score := 0

	// Track which lines we've scored to avoid double-counting
	// A line is identified by its direction and first tile position
	scoredLines := make(map[string]bool)

	for _, p := range placements {
		// Score horizontal line
		hLine := GetHorizontalLine(board, p.Pos)
		if len(hLine) > 1 {
			// Create unique key for this line
			// We use the first tile's position and direction
			key := "h" + hLine[0].String()
			if !scoredLines[key] {
				scoredLines[key] = true
				score += len(hLine)
				// Qwirkle bonus!
				if len(hLine) == 6 {
					score += 6
				}
			}
		}

		// Score vertical line
		vLine := GetVerticalLine(board, p.Pos)
		if len(vLine) > 1 {
			key := "v" + vLine[0].String()
			if !scoredLines[key] {
				scoredLines[key] = true
				score += len(vLine)
				if len(vLine) == 6 {
					score += 6
				}
			}
		}
	}

	// Single tile with no line neighbors scores 1 point
	if score == 0 && len(placements) == 1 {
		score = 1
	}

	return score
}

File: src/engine/rules.py
"""Rules engine: move validation for Qwirkle.

A valid move must:
1. Place tiles in a single row OR single column
2. Tiles must be contiguous (no gaps in the line)
3. Must connect to existing tiles (except first move)
4. Each resulting line must have ≤6 tiles with no duplicates
5. Each line must share exactly one attribute (all same color OR all same shape)
"""

from typing import List, Tuple, Set, Optional
from src.models.board import Board, Position
from src.models.tile import Tile, Color, Shape


def get_line_horizontal(board: Board, pos: Position) -> List[Tuple[Position, Tile]]:
    """Get the complete horizontal line containing a position.

    Extends left and right from pos until hitting empty cells.
    Includes the tile at pos if present.

    Args:
        board: The game board.
        pos: Starting position (row, col).

    Returns:
        List of (position, tile) tuples in left-to-right order.
    """
    row, col = pos
    tiles = []

    # Find leftmost tile in this line
    left = col
    while board.get((row, left - 1)) is not None:
        left -= 1

    # Collect all tiles from left to right
    current = left
    while True:
        tile = board.get((row, current))
        if tile is None:
            break
        tiles.append(((row, current), tile))
        current += 1

    return tiles


def get_line_vertical(board: Board, pos: Position) -> List[Tuple[Position, Tile]]:
    """Get the complete vertical line containing a position.

    Extends up and down from pos until hitting empty cells.
    Includes the tile at pos if present.

    Args:
        board: The game board.
        pos: Starting position (row, col).

    Returns:
        List of (position, tile) tuples in top-to-bottom order.
    """
    row, col = pos
    tiles = []

    # Find topmost tile in this line
    top = row
    while board.get((top - 1, col)) is not None:
        top -= 1

    # Collect all tiles from top to bottom
    current = top
    while True:
        tile = board.get((current, col))
        if tile is None:
            break
        tiles.append(((current, col), tile))
        current += 1

    return tiles


def is_valid_line(tiles: List[Tile]) -> bool:
    """Check if a line of tiles is valid.

    A valid line:
    - Has at most 6 tiles
    - Has no duplicate tiles
    - All tiles share exactly one attribute (all same color OR all same shape)

    Args:
        tiles: List of tiles in the line.

    Returns:
        True if the line is valid.
    """
    if len(tiles) == 0:
        return True  # Empty line is trivially valid

    if len(tiles) == 1:
        return True  # Single tile is always valid

    if len(tiles) > 6:
        return False  # Line too long

    # Check for duplicates
    if len(tiles) != len(set(tiles)):
        return False

    # Check shared attribute: all same color OR all same shape (not both, not neither)
    colors = {t.color for t in tiles}
    shapes = {t.shape for t in tiles}

    same_color = len(colors) == 1
    same_shape = len(shapes) == 1

    # Valid if exactly one attribute is shared
    # (same_color XOR same_shape, but also accept both for single tile which we already handled)
    if same_color and not same_shape:
        return True  # All same color, different shapes
    if same_shape and not same_color:
        return True  # All same shape, different colors

    return False


def are_positions_collinear(positions: List[Position]) -> Optional[str]:
    """Check if positions are all in the same row or column.

    Args:
        positions: List of (row, col) positions.

    Returns:
        'row' if all in same row, 'col' if all in same column,
        None if neither (or empty/single position returns 'row').
    """
    if len(positions) <= 1:
        return 'row'  # Trivially collinear

    rows = {p[0] for p in positions}
    cols = {p[1] for p in positions}

    if len(rows) == 1:
        return 'row'
    if len(cols) == 1:
        return 'col'
    return None


def are_positions_contiguous(positions: List[Position], direction: str) -> bool:
    """Check if positions form a contiguous line (no gaps).

    Args:
        positions: List of (row, col) positions.
        direction: 'row' (horizontal) or 'col' (vertical).

    Returns:
        True if positions are contiguous.
    """
    if len(positions) <= 1:
        return True

    if direction == 'row':
        # Same row, check columns are consecutive
        cols = sorted(p[1] for p in positions)
        for i in range(1, len(cols)):
            if cols[i] - cols[i-1] != 1:
                return False
    else:
        # Same column, check rows are consecutive
        rows = sorted(p[0] for p in positions)
        for i in range(1, len(rows)):
            if rows[i] - rows[i-1] != 1:
                return False

    return True


def get_affected_lines(board: Board, positions: List[Position]) -> List[List[Tuple[Position, Tile]]]:
    """Get all lines affected by placing tiles at the given positions.

    For each placed tile, we check its horizontal and vertical lines.
    Only returns lines with 2+ tiles (single tiles don't score).

    Args:
        board: The board with tiles already placed.
        positions: Positions where tiles were placed.

    Returns:
        List of lines, each line is a list of (position, tile) tuples.
    """
    seen_lines: Set[Tuple[Tuple[Position, Tile], ...]] = set()
    lines = []

    for pos in positions:
        # Check horizontal line
        h_line = get_line_horizontal(board, pos)
        if len(h_line) >= 2:
            key = tuple(h_line)
            if key not in seen_lines:
                seen_lines.add(key)
                lines.append(h_line)

        # Check vertical line
        v_line = get_line_vertical(board, pos)
        if len(v_line) >= 2:
            key = tuple(v_line)
            if key not in seen_lines:
                seen_lines.add(key)
                lines.append(v_line)

    return lines


def validate_move(
    board: Board,
    placements: List[Tuple[Position, Tile]],
    is_first_move: bool = False
) -> Tuple[bool, str]:
    """Validate a move before applying it.

    Args:
        board: Current board state (before placing tiles).
        placements: List of (position, tile) to place.
        is_first_move: True if this is the first move of the game.

    Returns:
        Tuple of (is_valid, error_message). If valid, error_message is empty.
    """
    if not placements:
        return False, "Must place at least one tile"

    positions = [p[0] for p in placements]
    tiles = [p[1] for p in placements]

    # Check all positions are empty
    for pos in positions:
        if board.get(pos) is not None:
            return False, f"Position {pos} is already occupied"

    # Check positions are collinear
    direction = are_positions_collinear(positions)
    if direction is None:
        return False, "All tiles must be placed in the same row or column"

    # Create a temporary board to check the result
    temp_board = board.copy()
    for pos, tile in placements:
        temp_board.place(pos, tile)

    # Check tiles are contiguous (including existing tiles in the line)
    # Get the full line after placement - starting from any placed position
    if direction == 'row':
        full_line = get_line_horizontal(temp_board, positions[0])
    else:
        full_line = get_line_vertical(temp_board, positions[0])

    line_positions = [p[0] for p in full_line]

    # All placed positions must be in the same contiguous line
    # If any placement position is not in the line, there's a gap
    for pos in positions:
        if pos not in line_positions:
            return False, "Tiles must be placed in a contiguous line"

    if not are_positions_contiguous(line_positions, direction):
        return False, "Tiles must be placed in a contiguous line"

    # Check connection to existing tiles (except first move)
    if not is_first_move:
        has_connection = False
        for pos in positions:
            # Check if any neighbor was on the original board
            for neighbor in temp_board.neighbor_positions(pos):
                if neighbor not in positions and board.get(neighbor) is not None:
                    has_connection = True
                    break
            if has_connection:
                break

        if not has_connection:
            return False, "Tiles must connect to existing tiles on the board"

    # Check all affected lines are valid
    affected = get_affected_lines(temp_board, positions)

    # For single tile placements that don't form lines with existing tiles,
    # we still need to validate the tile itself connects properly
    if not affected and len(placements) == 1 and not is_first_move:
        # This case means a single tile was placed but doesn't form a 2+ tile line
        # This is actually valid - the tile just doesn't score
        pass

    for line in affected:
        line_tiles = [t for _, t in line]
        if not is_valid_line(line_tiles):
            return False, f"Invalid line: tiles must share exactly one attribute with no duplicates"

    # Also check the main placement line even if it's not in affected
    # (in case it's a single-tile first move)
    if is_first_move and len(placements) > 1:
        if not is_valid_line(tiles):
            return False, "Invalid line: tiles must share exactly one attribute with no duplicates"

    return True, ""
